core/src/main/java/org/apache/iceberg/UpdateSnapshotReferencesOperation.java

```diff
@@ -0,0 +1,166 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.iceberg.exceptions.CommitFailedException;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.util.Tasks;
+
+import static org.apache.iceberg.TableProperties.COMMIT_MAX_RETRY_WAIT_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_MAX_RETRY_WAIT_MS_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_MIN_RETRY_WAIT_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_MIN_RETRY_WAIT_MS_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_NUM_RETRIES;
+import static org.apache.iceberg.TableProperties.COMMIT_NUM_RETRIES_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_TOTAL_RETRY_TIME_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_TOTAL_RETRY_TIME_MS_DEFAULT;
+
+class UpdateSnapshotReferencesOperation implements PendingUpdate<Map<String, SnapshotRef>> {
+
+  private final TableOperations ops;
+  private TableMetadata base;
+  private final Map<String, SnapshotRef> refsToUpdate;
+  private final Set<String> refsToRemove;
+
+  UpdateSnapshotReferencesOperation(TableOperations ops) {
+    this.ops = ops;
+    this.base = ops.current();
+    this.refsToUpdate = Maps.newHashMap();
+    this.refsToRemove = Sets.newHashSet();
+  }
+
+  @Override
+  public Map<String, SnapshotRef> apply() {
+    this.base = ops.refresh();
+    Map<String, SnapshotRef> refs = Maps.newHashMap();
+    TableMetadata.Builder updatedBuilder = TableMetadata.buildFrom(base);
+    for (String refName : refsToRemove) {
+      SnapshotRef ref = base.ref(refName);
+      ValidationException.check(ref != null, "Cannot remove nonexistent snapshot ref %s", refName);
+      if (ref.isBranch()) {
+        updatedBuilder.removeBranch(refName);
+      }
+      else {
+        updatedBuilder.removeTag(refName);
+      }
+    }
+    for (Map.Entry<String, SnapshotRef> refEntry : refsToUpdate.entrySet()) {
+      String name = refEntry.getKey();
+      SnapshotRef snapshotRef = refEntry.getValue();
+      updatedBuilder.createSnapshotRef(name, snapshotRef);
+    }
+    return refs;
+  }
+
+  @Override
+  public void commit() {
+    Tasks.foreach(ops)
+        .retry(base.propertyAsInt(COMMIT_NUM_RETRIES, COMMIT_NUM_RETRIES_DEFAULT))
```

Comment:
I don't think this operation should retry, like SchemaUpdate. Updating a schema applies changes to the current schema, just like this applies changes to the current set of refs. In order to retry, this would need to keep track of the changes and re-apply them each time it attempts to commit. Since this isn't tracking changes (and would be harder to do that way) I suggest just not retrying.

Reply:
Yes, makes sense. Updated to remove retrying